include/dlaf/eigensolver/tridiag_solver/mc.h

```diff
@@ -203,53 +197,49 @@ void offloadDiagonal(Matrix<const T, Device::CPU>& mat_trd, Matrix<T, Device::CP
+                          Matrix<SizeType, Device::CPU>(vec_size, vec_tile_size),  // i2
+                          Matrix<SizeType, Device::CPU>(vec_size, vec_tile_size),  // i3
+                          Matrix<ColType, Device::CPU>(vec_size, vec_tile_size)};  // c
+
+  matrix::MatrixMirror<T, Device::CPU, device> tridiag_h(mat_trd);
+
+  WorkSpaceHostMirror<T, device> ws_h{matrix::MatrixMirror<T, Device::CPU, device>(evals),     // evals
+                                      matrix::MatrixMirror<T, Device::CPU, device>(evecs),     // evecs
+                                      matrix::MatrixMirror<T, Device::CPU, device>(ws.mat1),   // mat1
+                                      matrix::MatrixMirror<T, Device::CPU, device>(ws.mat2)};  // mat2
+
   // Set `mat_ev` to `zero` (needed for Given's rotation to make sure no random values are picked up)
-  matrix::util::set0<Backend::MC, T, Device::CPU>(pika::execution::thread_priority::normal, mat_ev);
+  matrix::util::set0<backend, T, device>(pika::execution::thread_priority::normal, evecs);
+  ws_h.evecs.copySourceToTarget();  // copy from GPU to CPU if evecs is on GPU
 
   // Cuppen's decomposition
-  std::vector<pika::shared_future<T>> offdiag_vals = cuppensDecomposition(mat_trd);
+  std::vector<pika::shared_future<T>> offdiag_vals = cuppensDecomposition(tridiag_h.get());
 
   // Solve with stedc for each tile of `mat_trd` (nb x 2) and save eigenvectors in diagonal tiles of
   // `mat_ev` (nb x nb)
-  solveLeaf(mat_trd, mat_ev);
+  solveLeaf(tridiag_h.get(), ws_h.evecs.get());
 
   // Offload the diagonal from `mat_trd` to `d`
-  offloadDiagonal(mat_trd, d);
-
-  // Auxiliary matrix used for the D&C algorithm
-  const matrix::Distribution& distr = mat_ev.distribution();
-
-  LocalElementSize vec_size(distr.size().rows(), 1);
-  TileElementSize vec_tile_size(distr.blockSize().rows(), 1);
-  WorkSpace<T> ws{Matrix<T, Device::CPU>(distr),
-                  Matrix<T, Device::CPU>(distr),
-                  Matrix<T, Device::CPU>(vec_size, vec_tile_size),
-                  Matrix<T, Device::CPU>(vec_size, vec_tile_size),
-                  Matrix<T, Device::CPU>(vec_size, vec_tile_size),
-                  Matrix<SizeType, Device::CPU>(vec_size, vec_tile_size),
-                  Matrix<SizeType, Device::CPU>(vec_size, vec_tile_size),
-                  Matrix<SizeType, Device::CPU>(vec_size, vec_tile_size),
-                  Matrix<ColType, Device::CPU>(vec_size, vec_tile_size)};
+  offloadDiagonal(tridiag_h.get(), ws_h.evals.get());
 
   // Each triad represents two subproblems to be merged
   for (auto [i_begin, i_split, i_end] : generateSubproblemIndices(distr.nrTiles().rows())) {
-    mergeSubproblems(i_begin, i_split, i_end, ws, offdiag_vals[to_sizet(i_split)], d, mat_ev);
+    mergeSubproblems<backend>(i_begin, i_split, i_end, offdiag_vals[to_sizet(i_split)], ws, ws_h, evals,
+                              evecs);
```

Comment:
Do `evals` and `evecs` not need to be copied back to the source before doing this? Possibly by scoping `ws_h` to go out of scope before reaching this for loop? Or is the synchronization done somewhere else implicitly?

Reply:
Hopefully answering my own question after reading further: the synchronization is done inside `mergeSubproblems`.